flask_ansible.md

```diff
@@ -0,0 +1,182 @@
+
+Some alternatives to do the task of provisioning the code to the remote machine would be to use the synchronize module (based on rsync) to keep directories in sync or even the git module to check out code from a repository on the remote server.
+
+### Installing necessary libraries
+```
+  - name: Install extra packages (to be able to install pip)
+    yum: name=epel-release state=latest
+    sudo: yes
+
+  - name: Install lsof (for better process management)
+    yum: name=lsof state=latest
+    sudo: yes
+
+  - name: Install pip
+    yum: name=python-pip state=latest
+    sudo: yes
+
+  - name: Install virtualenv
+    pip: name=virtualenv
+    sudo: yes
+
+  - name: Create virtualenv for project
+    shell: virtualenv "{{ PROJECT_HOME }}/venv"
+           creates="{{ PROJECT_HOME }}/venv/bin/activate"
+
+  - name: Install Flask in the virtualenv 
+    pip: name=flask virtualenv="{{ PROJECT_HOME }}/venv" 
+```
+
+The first task installs the packages necessary to install pip, Python's package manager. We also install lsof to use it later to easily terminate and restart our website process. These actions require sudo privileges, so we specify that as well. We use the yum module, that is, Ansible's module for the CentOS package manager. Let's say we tried to use the shell module instead, ie something like:
+```
+- name : Install epel
+  shell: yum install epel-release
+```
+Then, Ansible would in fact get stuck because yum requires the user to confirm that they want to install the packages by using the keyboard. We could of course specify the -y flag to force the confirmation, but in general there's no need to use the shell when a well-documented ansible core module will do.
+
+
+
+For the next task, we install virtualenv to separate the packages used for the website from the system Python. For such a simple example this is not really needed, but when deploying at a server with several python projects installed, separating each one in its own virtual environment makes life easier. 
+
+The next task (creating a virtual environment for our project) is quite interesting from an indempotency point of view. We want to be able to know if the task has ran already, so that Ansible skips it the second time. By defining that this task creates a particular file when it has ran, Ansible knows to skip the whole task if the file already exists. This logic of course could be wrapped more neatly in a dedicated virtualenv module, but at the time of this writing, such a module doesn't exist in the core modules of Ansible.
+
+Finally, we install the Flask library in the virtual environment we just created within our project directory.
+
+
+### Launching the website
+```
+  - name: Get process id running the webserver
+    shell: lsof -t -i:{{PORT}}
+    sudo: yes
+    ignore_errors: yes
+    register: pid
+
+  - name: Kill webserver if it's already running
+    shell: kill {{pid.stdout}}
+    when: pid.stdout != ""
+
```

Comment:
here it can happen that if the bourne shell is not used, source is not found and the thing waits for all eternity